functions/ts/src/sops.ts

```diff
@@ -0,0 +1,206 @@
+import {
+  Configs,
+  KubernetesObject,
+  isKubernetesObject,
+  generalResult,
+  getAnnotation,
+  addAnnotation,
+  removeAnnotation,
+} from 'kpt-functions';
+import { DumpOptions, safeDump, safeLoad } from 'js-yaml';
+import rw from 'rw';
+import { spawnSync, execSync } from 'child_process';
+
+const TEMP_PATH = '/tmp/tmp.yaml';
+
+const IGNORE_MAC = 'ignore-mac';
+const VERBOSE = 'verbose';
+const OVERRIDE_PREEXEC_CMD = 'override-preexec-cmd';
+
+// pre-exec command may be ovverriden from config
+let preExecCmd =
+  '[ "$SOPS_IMPORT_PGP" == "" ] || (echo "$SOPS_IMPORT_PGP" | gpg --import)';
+
+interface SopsKubernetesObject extends KubernetesObject {}
+
+function isSopsKubernetesObject(
+  o: KubernetesObject
+): o is SopsKubernetesObject {
+  return o && o.hasOwnProperty('sops');
+}
+
+const YAML_STYLE: DumpOptions = {
+  // indentation width to use (in spaces).
+  indent: 2,
+  // when true, will not add an indentation level to array elements.
+  noArrayIndent: true,
+  // 'undefined' is an invalid value for safeDump.
+  // TODO(frankf): Explore ways to make this safer.
+  // Either reason about not having 'undefined' in all cases OR
+  // only skip 'undefined'.
+  skipInvalid: true,
+  // unset lineWidth from default of 80 to avoid reformatting
+  lineWidth: -1,
+  // avoid refs because many YAML parsers in the k8s ecosystem don't support them
+  noRefs: true,
+};
+
+async function writeFile(path: string, data: string): Promise<void> {
+  return new Promise((resolve, reject) => {
+    rw.writeFile(path, data, 'utf8', (err: object) => {
+      if (err) return reject(err);
+      resolve();
+    });
+  });
+}
+
+export async function sops(configs: Configs) {
+  // Validate config data and read arguments.
+  const args = readSopsArguments(configs);
+
+  // run preexec if it's not empty
+  if (preExecCmd !== '') {
+    execSync(preExecCmd);
+  }
+
+  // Put all documents into the local place and
+  // cleanup the configs storage. This is to keep
+  // the order of documents
+  const allDocs = configs.getAll();
+  configs.deleteAll();
+
+  for (const object of allDocs) {
+    if (isSopsKubernetesObject(object)) {
+      // this function
+      await decryptAndInsertSopsKubernetesObject(args, configs, object);
+    } else {
+      // add it back
+      configs.insert(object);
+    }
+  }
+}
+
+function detachConfigKubernetesIoAnnotations(
+  object: SopsKubernetesObject
+): Map<string, string> {
+  const annotations: string[] = [
+    'config.kubernetes.io/index',
+    'config.kubernetes.io/path',
+    'config.k8s.io/id',
+    'kustomize.config.k8s.io/id',
```

Comment:
Why not just remove any annotations with prefix `config.kubernetes.io`?


@monopole What's `kustomize.config.k8s.io/id` and how's it different from `config.k8s.io/id`?
Same with `config.kubernetes.io` vs `config.k8s.io`. Why do these different prefixes co-exist?

Reply:
first 3 used by kustomize/kpt fn run. 
the last one used by kustomize build to check keep information about the document.

>Why not just remove any annotations with prefix config.kubernetes.io?


hmm. really good catch. that makes total sense if we're confident that all annotations with that prefixes are added by kustomize/kpt and were not part of the original documents. waiting for **monopole** to chime it

Reply:
@frankfarzan , I've checked -
There are at least 3 more annotations with the same prefixes that we don't want to remove accidentally:

- config.kubernetes.io/local-config
- config.kubernetes.io/function
- config.k8s.io/function

I think for now it's better to keep this implementation. If we're not sure what will happen with this annotations I can add a parameter to override this list in the function configuration.

Reply:
done

Reply:
Filed an issue: https://github.com/kubernetes-sigs/kustomize/issues/2996

Reply:
Let's add a TODO to this issue.

Reply:
I don't think there's a canonical list of well-known annotation keys or `apiVersion` values, or policies around using the latter as the former, that one can reliably use for any given goal.  So providing a means to specify such a "detach or ignore" list, and offering a good guess as a default, looks good to me.

e.g. this is interesting, but incomplete :  https://kubernetes.io/docs/reference/kubernetes-api/labels-annotations-taints

Reply:
I was thinking about the same: created the list of annotations that should be detached before the decryption, but I also added a parameter called [override-detached-annotations](https://github.com/GoogleContainerTools/kpt-functions-catalog/pull/101/files#diff-11b1a6e5f9cc9e2bb106e55858164b37R199) that allows to specify comma-separated list of annotations to be detached if the behavior should be changed for any reason.